File: packages/liveblocks-node/src/Session.ts
import type {
  IUserInfo,
  Json,
  JsonObject,
  URLSafeString,
} from "@liveblocks/core";
import { normalizeRoomPermissions, Permission, url } from "@liveblocks/core";

import type { AuthResponse } from "./client";
import { assertNonEmpty, normalizeStatusCode } from "./utils";

const MAX_PERMS_PER_SET = 10;

/**
 * Assign this to a room (or wildcard pattern) if you want to grant the user
 * read permissions to the room. (Note that the user will still have permissions
 * to update their own presence.)
 */
const READ_ACCESS = Object.freeze([Permission.Read] as const);

/**
 * Assign this to a room (or wildcard pattern) if you want to grant the user
 * permissions to read and write to the room.
 */
const FULL_ACCESS = Object.freeze([Permission.Write] as const);

const roomPatternRegex = /^([*]|[^*]{1,128}[*]?)$/;

type PostFn = (path: URLSafeString, json: Json) => Promise<Response>;

/**
 * Class to help you construct the exact permission set to grant a user.
 *
 * Usage:
 *
 *    const session = liveblocks.prepareSession();
 *    session.allow(roomId, permissions)
 *
 * For the `permissions` argument, pass a list of permission scopes.
 *
 *    session.allow('my-room', ['*:write'])   // Read + write access
 *    session.allow('my-room', ['*:read'])    // Read-only access
 *    session.allow('my-room', [
 *      '*:write',                            // Read + write access by default
 *      'comments:read'                       // But read-only access to comments
 *      'feeds:none'                          // And no access to feeds
 *    ])
 *
 * Rooms can be specified with a prefix match, if the name ends in an asterisk.
 * In that case, access is granted to *all* rooms that start with that prefix:
 *
 *    // Read + write access to *all* rooms that start with "abc:"
 *    session.allow('abc:*', ['*:write'])
 *
 * You can define at most 10 room IDs (or patterns) in a single token,
 * otherwise the token would become too large and unwieldy.
 *
 */
export class Session {
  /**
   * @deprecated Use `["*:write"]` instead.
   */
  public readonly FULL_ACCESS = FULL_ACCESS;

  /**
   * @deprecated Use `["*:read"]` instead.
   */
  public readonly READ_ACCESS = READ_ACCESS;

  #postFn: PostFn;
  #userId: string;
  #userInfo?: IUserInfo;
  #organizationId?: string;
  /** Only used as a hint to produce better error messages. */
  #localDev: boolean;
  #sealed = false;
  readonly #permissions: Map<string, Set<Permission>> = new Map();

  /** @internal */
  constructor(
    postFn: PostFn,
    userId: string,
    userInfo?: IUserInfo,
    organizationId?: string,
    localDev?: boolean
  ) {
    assertNonEmpty(userId, "userId"); // TODO: Check if this is a legal userId value too

    this.#postFn = postFn;
    this.#userId = userId;
    this.#userInfo = userInfo;
    this.#organizationId = organizationId;
    this.#localDev = localDev ?? false;
  }

  #getOrCreate(roomId: string): Set<Permission> {
    if (this.#sealed) {
      throw new Error("You can no longer change these permissions.");
    }

    let perms = this.#permissions.get(roomId);
    if (perms) {
      return perms;
    } else {
      if (this.#permissions.size >= MAX_PERMS_PER_SET) {
        throw new Error(
          "You cannot add permissions for more than 10 rooms in a single token"
        );
      }

      perms = new Set<Permission>();
      this.#permissions.set(roomId, perms);
      return perms;
    }
  }

  public allow(roomIdOrPattern: string, newPerms: readonly Permission[]): this {
    if (typeof roomIdOrPattern !== "string") {
      throw new Error("Room name or pattern must be a string");
    }
    if (!roomPatternRegex.test(roomIdOrPattern)) {
      throw new Error("Invalid room name or pattern");
    }

    if (newPerms.length === 0) {
      throw new Error("Permission list cannot be empty");
    }

    const permissions = normalizeRoomPermissions(newPerms);

    const existingPerms = this.#getOrCreate(roomIdOrPattern);
    for (const perm of permissions) {
      existingPerms.add(perm);
    }
    return this; // To allow chaining multiple allow calls
  }

  /** @internal - For unit tests only */
  public hasPermissions(): boolean {
    return this.#permissions.size > 0;
  }

  /** @internal - For unit tests only */
  public seal(): void {
    if (this.#sealed) {
      throw new Error(
        "You cannot reuse Session instances. Please create a new session every time."
      );
    }
    this.#sealed = true;
  }

  /** @internal - For unit tests only */
  public serializePermissions(): JsonObject {
    return Object.fromEntries(
      Array.from(this.#permissions.entries()).map(([pat, perms]) => [
        pat,
        Array.from(perms),
      ])
    );
  }

  /**
   * Call this to authorize the session to access Liveblocks. Note that this
   * will return a Liveblocks "access token". Anyone that obtains such access
   * token will have access to the allowed resources.
   */
  public async authorize(): Promise<AuthResponse> {
    this.seal();
    if (!this.hasPermissions()) {
      console.warn(
        "Access tokens without any permission will not be supported soon, you should use wildcards when the client requests a token for resources outside a room. See https://liveblocks.io/docs/errors/liveblocks-client/access-tokens-not-enough-permissions"
      );
    }

    try {
      const body: {
        userId: string;
        permissions: JsonObject;
        userInfo?: IUserInfo;
        organizationId?: string;
      } = {
        // Required
        userId: this.#userId,
        permissions: this.serializePermissions(),

        // Optional metadata
        userInfo: this.#userInfo,
      };

      if (this.#organizationId !== undefined) {
        body.organizationId = this.#organizationId;
      }

      const resp = await this.#postFn(url`/v2/authorize-user`, body);

      return {
        status: normalizeStatusCode(resp.status),
        body: await resp.text(),
      };
    } catch (er) {
      return {
        status: 503 /* Service Unavailable */,
        body: this.#localDev
          ? "Could not connect to your Liveblocks dev server. Is it running?"
          : 'Call to /v2/authorize-user failed. See "error" for more information.',
        error: er as Error | undefined,
      };
    }
  }
}

File: packages/liveblocks-python/liveblocks/session.py
from __future__ import annotations

import re
import warnings
from collections.abc import Sequence
from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from liveblocks.client import AsyncLiveblocks, Liveblocks
    from liveblocks.models.authorize_user_request_body import AuthorizeUserRequestBody
    from liveblocks.models.authorize_user_response import AuthorizeUserResponse

Permission = Literal[
    "room:write",
    "room:read",
    "room:presence:write",
    "comments:write",
    "comments:read",
]

ALL_PERMISSIONS: frozenset[str] = frozenset(
    ["room:write", "room:read", "room:presence:write", "comments:write", "comments:read"]
)

READ_ACCESS: tuple[Permission, ...] = ("room:read", "room:presence:write", "comments:read")
FULL_ACCESS: tuple[Permission, ...] = ("room:write", "comments:write")

_MAX_PERMS_PER_SET = 10
_ROOM_PATTERN_RE = re.compile(r"^([*]|[^*]{1,128}[*]?)$")


class _BaseSession:
    """Shared permission-building logic for sync and async sessions."""

    FULL_ACCESS = FULL_ACCESS
    READ_ACCESS = READ_ACCESS

    def __init__(
        self,
        user_id: str,
        user_info: dict[str, Any] | None = None,
        organization_id: str | None = None,
    ) -> None:
        if not user_id:
            raise ValueError(
                "Invalid value for 'user_id'. Please provide a non-empty string. "
                "For more information: https://liveblocks.io/docs/api-reference/liveblocks-node#authorize"
            )

        self._user_id = user_id
        self._user_info = user_info
        self._organization_id = organization_id
        self._sealed = False
        self._permissions: dict[str, list[str]] = {}

    def _get_or_create(self, room_id: str) -> list[str]:
        if self._sealed:
            raise RuntimeError("You can no longer change these permissions.")

        perms = self._permissions.get(room_id)
        if perms is not None:
            return perms

        if len(self._permissions) >= _MAX_PERMS_PER_SET:
            raise RuntimeError("You cannot add permissions for more than 10 rooms in a single token")

        perms = []
        self._permissions[room_id] = perms
        return perms

    def allow(self, room_id_or_pattern: str, permissions: Sequence[Permission]) -> _BaseSession:
        if not _ROOM_PATTERN_RE.match(room_id_or_pattern):
            raise ValueError("Invalid room name or pattern")
        if not permissions:
            raise ValueError("Permission list cannot be empty")

        existing = self._get_or_create(room_id_or_pattern)
        for perm in permissions:
            if perm not in ALL_PERMISSIONS:
                raise ValueError(f"Not a valid permission: {perm}")
            if perm not in existing:
                existing.append(perm)
        return self

    @property
    def _has_permissions(self) -> bool:
        return bool(self._permissions)

    def _serialize_permissions(self) -> dict[str, list[str]]:
        return {pat: list(perms) for pat, perms in self._permissions.items()}

    def _build_request_body(self) -> AuthorizeUserRequestBody:
        from liveblocks.models.authorize_user_request_body import AuthorizeUserRequestBody
        from liveblocks.models.authorize_user_request_body_permissions import (
            AuthorizeUserRequestBodyPermissions,
        )
        from liveblocks.models.authorize_user_request_body_user_info import (
            AuthorizeUserRequestBodyUserInfo,
        )

        if self._sealed:
            raise RuntimeError("You cannot reuse Session instances. Please create a new session every time.")
        self._sealed = True
        if not self._permissions:
            warnings.warn(
                "Access tokens without any permission will not be supported soon, "
                "you should use wildcards when the client requests a token for "
                "resources outside a room. See https://liveblocks.io/docs/errors/"
                "liveblocks-client/access-tokens-not-enough-permissions",
                stacklevel=3,
            )

        perms_model = AuthorizeUserRequestBodyPermissions()
        perms_model.additional_properties = self._serialize_permissions()

        user_info_model: AuthorizeUserRequestBodyUserInfo | None = None
        if self._user_info is not None:
            user_info_model = AuthorizeUserRequestBodyUserInfo()
            user_info_model.additional_properties = dict(self._user_info)

        body = AuthorizeUserRequestBody(
            user_id=self._user_id,
            permissions=perms_model,
        )

        if user_info_model is not None:
            body.user_info = user_info_model
        if self._organization_id is not None:
            body.organization_id = self._organization_id

        return body


class Session(_BaseSession):
    """Synchronous session. Created by ``Liveblocks.prepare_session()``.

    Usage::

        session = client.prepare_session("user-123")
        session.allow("my-room", session.FULL_ACCESS)
        result = session.authorize()
    """

    def __init__(
        self,
        client: Liveblocks,
        user_id: str,
        user_info: dict[str, Any] | None = None,
        organization_id: str | None = None,
    ) -> None:
        super().__init__(user_id, user_info=user_info, organization_id=organization_id)
        self._client = client

    def allow(self, room_id_or_pattern: str, permissions: Sequence[Permission]) -> Session:
        super().allow(room_id_or_pattern, permissions)
        return self

    def authorize(self) -> AuthorizeUserResponse:
        body = self._build_request_body()
        return self._client.authorize_user(body=body)


class AsyncSession(_BaseSession):
    """Asynchronous session. Created by ``AsyncLiveblocks.prepare_session()``.

    Usage::

        session = client.prepare_session("user-123")
        session.allow("my-room", session.FULL_ACCESS)
        result = await session.authorize()
    """

    def __init__(
        self,
        client: AsyncLiveblocks,
        user_id: str,
        user_info: dict[str, Any] | None = None,
        organization_id: str | None = None,
    ) -> None:
        super().__init__(user_id, user_info=user_info, organization_id=organization_id)
        self._client = client

    def allow(self, room_id_or_pattern: str, permissions: Sequence[Permission]) -> AsyncSession:
        super().allow(room_id_or_pattern, permissions)
        return self

    async def authorize(self) -> AuthorizeUserResponse:
        body = self._build_request_body()
        return await self._client.authorize_user(body=body)
